src/core/ucc_team.c

```diff
@@ -0,0 +1,216 @@
+    memcpy(&b_params.params, &team->params, sizeof(ucc_team_params_t));
+    team->cl_teams = ucc_malloc(sizeof(ucc_cl_team_t *) * context->n_cl_ctx);
+    if (!team) {
+        ucc_error("failed to allocate %zd bytes for cl teams array",
+                  sizeof(ucc_cl_team_t *) * context->n_cl_ctx);
+        return UCC_ERR_NO_MEMORY;
+    }
+    for (i = 0; i < context->n_cl_ctx; i++) {
+        cl_iface = UCC_CL_CTX_IFACE(context->cl_ctx[i]);
+        status   = cl_iface->team.create_post(&context->cl_ctx[i]->super,
+                                            &b_params, &b_team);
+
+        if (status != UCC_OK) {
+            ucc_debug("failed to create CL %s team", cl_iface->super.name);
+            /* TODO: what is the logic here? continue or report error?
+               maybe need to recheck supported matrix after the whole
+               ucc_team is created and report error if not all requested
+               capabilities are provided */
+            continue;
+        }
+        status = cl_iface->team.create_test(b_team);
+        if (status < 0) {
+            ucc_debug("failed to create CL %s team", cl_iface->super.name);
+            /* TODO: see comment above */
+            continue;
+        }
+        team->cl_teams[team->n_cl_teams++] =
+            ucc_derived_of(b_team, ucc_cl_team_t);
+        if (status == UCC_INPROGRESS) {
+            /* workaround to fix oob allgather issue if multiple teams use it
+               simultaneously*/
+            break;
+        }
+    }
+    if (team->n_cl_teams == 0) {
+        ucc_error("no cl teams have been created for ucc_team");
+        return UCC_ERR_NO_MESSAGE;
+    }
+    team->last_team_create_posted = i;
+    team->status = (i < context->n_cl_ctx) ? UCC_INPROGRESS : UCC_OK;
+    return UCC_OK;
+}
+
+ucc_status_t ucc_team_create_post(ucc_context_h *contexts, uint32_t num_contexts,
+                                  const ucc_team_params_t *params,
+                                  ucc_team_h *new_team)
+{
+    ucc_team_t  *team;
+    ucc_status_t status;
+    if (num_contexts < 1) {
+        return UCC_ERR_INVALID_PARAM;
+    } else if (num_contexts > 1) {
+        ucc_error("team creation from multiple contexts is not supported yet");
+        return UCC_ERR_NOT_SUPPORTED;
+    }
+    team = ucc_malloc(sizeof(ucc_team_t), "ucc_team");
+    if (!team) {
+        ucc_error("failed to allocate %zd bytes for ucc team",
+                  sizeof(ucc_team_t));
+        return UCC_ERR_NO_MEMORY;
+    }
+    team->n_cl_teams   = 0;
+    team->num_contexts = num_contexts;
+    team->contexts =
+        ucc_malloc(sizeof(ucc_context_t *) * num_contexts, "ucc_team_ctx");
+    if (!team->contexts) {
+        ucc_error("failed to allocate %zd bytes for ucc team contexts array",
+                  sizeof(ucc_context_t) * num_contexts);
+        status = UCC_ERR_NO_MEMORY;
+        goto err_ctx_alloc;
+    }
+    memcpy(team->contexts, contexts, sizeof(ucc_context_t *) * num_contexts);
```

Comment:
Though not performance critical, I was wondering why not use ucs_memcpy* ?

Reply:
why to use it ? do we want to overwrite default memcpy with ucc_memcpy throughout the code? we can

Reply:
In general for performance portability across the architectures and in this case for uniformity. As I understand, folks are always optimizing the performance in UCS.

Reply:
so, Manju, basically ucs does not have ucs_memcpy. There are 2 flavours of memcpy there: ucs_memcpy_nontemporal and ucs_memcpy_relaxed with arch specific implementations. And UCX itself use those only in a couple of places. All the rest of UCX code uses just memcpy - which is way more portable actually (libc on pretty much any system has memcpy). So, i don't think we should use some exotic memcpy in the places like above - out of critical path, generic copy w/o any specific semantics.